variable/variable.cpp

```diff
@@ -29,6 +27,45 @@ Variable::Variable(const Dimensions &dims, VariableConceptHandle data)
 Variable::Variable(const llnl::units::precise_measurement &m)
     : Variable(m.value() * units::Unit(m.units())) {}
 
+namespace {
+void check_nested_in_assign(const Variable &lhs, const Variable &rhs) {
+  if (!rhs.is_valid() || rhs.dtype() != dtype<Variable>) {
+    return;
+  }
+  // In principle we should also check when the RHS contains DataArrays or
+  // Datasets. But those are copied when stored in Variables,
+  // so no check needed here.
+  for (const auto &nested : rhs.values<Variable>()) {
+    if (&lhs == &nested) {
+      throw std::invalid_argument("Cannot assign Variable, the right hand side "
+                                  "contains a reference to the left hand side. "
+                                  "Reference cycles are not allowed.");
+    }
+    check_nested_in_assign(lhs, nested);
+  }
+}
+} // namespace
+
+Variable &Variable::operator=(const Variable &other) {
+  check_nested_in_assign(*this, other);
+  m_dims = other.m_dims;
+  m_strides = other.m_strides;
+  m_offset = other.m_offset;
+  m_object = other.m_object;
+  m_readonly = other.m_readonly;
+  return *this;
```

Comment:
```suggestion
  return *this = Variable(other);
```
?

Reply:
Possible. Do you trust the optimizer enough?

Reply:
Why optimizer? It is just using move semantics, which are guarenteed?

Reply:
Sure. But it is a copy + move instead of just a copy.

Reply:
True. Should be cheap though (compared to typical variable sizes)? And as you say, to optimizer might take care of it.